File: leetcode/cpp/PassedProblems/143-reorder-list.cpp
/*
 * 143-reorder-list.cpp
 *
 *  Created on: Jul 14, 2015
 *      Author: hanzhe
 */

#include <iostream>
using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
  };
class Solution {
public:
    void reorderList(ListNode* head) {
    	if(head == NULL || head->next == NULL)
    		return;
        /*1. find middle node*/
    	ListNode * pf = head, *ps = head;
    	while((pf!= NULL && pf->next != NULL && pf->next->next != NULL))
    	{
    		pf = pf->next->next;
    		ps = ps->next;
    	}

    	/*2. reverse sec half nodes */
    	ListNode * ah = ps->next;
    	ps->next = NULL;
    	ListNode * n = ah->next;
    	ah->next = NULL;
    	while(n != NULL)
    	{
    		ListNode * nn = n->next;
    		n->next = ah;
    		ah = n;
    		n = nn;
    	}

    	/*3. merge 2 list of nodes*/
    	ListNode * h1 = head;
    	while(ah != NULL)
    	{
    		ListNode* nah = ah->next;
    		ListNode* nh = h1->next;
    		ah->next = nh;
    		h1->next = ah;
    		ah = nah;
    		h1 = nh;
    	}
    }

    void test()
    {
    	ListNode * p1 = new ListNode(1), * p2 = new ListNode(2);
    	p1->next = p2;
    	reorderList(p1);
    	while(p1 != NULL)
    	{
    		cout << p1->val;
    		p1 = p1->next;
    	}
    }
};

File: leetcode/python/passedProblems/143-reorder-list.py
#!/usr/bin/env python
# encoding: utf-8

"""
Given a singly linked list L: L0→L1→…→Ln-1→Ln,
reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…

You must do this in-place without altering the nodes' values.

For example,
Given {1,2,3,4}, reorder it to {1,4,2,3}.
"""


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# 常规方法：split2段，后段reverse，2段合并
class Solution(object):
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: void Do not return anything, modify head in-place instead.
        """
        # 数组存贮位置, 不是好的方法（速度还行）
        if head == None: return
        ls = []
        tmp = head
        while tmp != None:
            ls.append(tmp)
            tmp = tmp.next
        for i in range(len(ls)/2):
            ls[i].next = ls[len(ls)-i-1]
            ls[len(ls)-i-1].next = ls[i+1]
        ls[len(ls)/2].next = None



